File: src/java/magma/rule/TypeRule.java
package magma.rule;

import magma.node.Node;
import magma.result.Err;
import magma.result.Ok;
import magma.result.Result;

/**
 * A Rule that attaches a type tag to nodes when lexing and checks for the type before generating.
 * This rule wraps another rule and adds/checks for a specific type tag.
 */
public final class TypeRule implements Rule {
    private final String type;
    private final Rule childRule;

    /**
     * Creates a new TypeRule with the specified type and child rule.
     *
     * @param type the type tag to attach/check for
     * @param childRule the rule to delegate to
     */
    public TypeRule(final String type, final Rule childRule) {
        this.type = type;
        this.childRule = childRule;
    }

    @Override
    public Result<String, String> generate(final Node node) {
        // Only generate if the node has the expected type
        if (!node.is(this.type)) {
            return new Err<>("Node does not have the expected type: " + this.type);
        }
        
        // Delegate to the child rule
        return this.childRule.generate(node);
    }

    @Override
    public Result<Node, String> lex(final String input) {
        // Delegate to the child rule
        final Result<Node, String> result = this.childRule.lex(input);
        
        // If the child rule successfully lexed the input, attach the type tag
        if (result.isErr()) {
            return result;
        }

        return new Ok<>(result.unwrap().retype(this.type));
    }
}

File: src/node/magma/rule/TypeRule.ts
/*import magma.error.CompileError;*/
/*import magma.node.Node;*/
/*import magma.result.Err;*/
/*import magma.result.Ok;*/
/*import magma.result.Result;*/
/*import java.util.Optional;*/
/*public record TypeRule(String type, Rule rule) implements Rule {
	private Optional<Node> lex0(final String input) {
		return this.rule.lex(input).findValue().map(node -> node.retype(this.type));
	}

	private Optional<String> generate0(final Node node) {
		if (node.is(this.type)) return this.rule.generate(node).findValue();
		return Optional.empty();
	}

	@Override
	public Result<Node, CompileError> lex(final String input) {
		return this.lex0(input).<Result<Node, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}

	@Override
	public Result<String, CompileError> generate(final Node node) {
		return this.generate0(node).<Result<String, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}
}*/
